File: medium/199.ts
/**
    199. Binary Tree Right Side View
 */

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function rightSideView(root: TreeNode | null): number[] {
    const res: number[] = []
    dfs(root, 0)
    return res

    function dfs(root: TreeNode, level: number) {
        if (root == null) return false
        res[level] = root.val
        dfs(root.left, level + 1)
        dfs(root.right, level + 1)
    }
};

File: medium/199.cpp
// 199. Binary Tree Right Side View

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        dfs(root, res, 0);
        return res;
    }

private:
    void dfs(TreeNode *root, vector<int>& res, int level){
        if(!root) return;
        if(res.size() < level + 1) res.push_back(root->val);
        else res[level] = root->val;
        // res[level] = root->val;
        dfs(root->left, res, level + 1);
        dfs(root->right, res, level + 1);
    }
};
